### DISCUSSION TITLE
Astro with Contextual Editing

### DISCUSSION THREAD

**Discussion by apolo-damasco** [tinacms/tinacms#3399] [Ideas]
Hi, I just wanted to start this request because I know that many people are interested in this and would like to have a way to track the news on this topic.

@perkinsjr I see on discord that you are looking into this. Would you mind writing here if you ever manage to make it work?

Thank you!

**Update 28/5/2026**
The Astro starter template has been updated to remove the react dependency.

**Update 24/3/2025**
Starter template with Astro+TinaCMS - https://github.com/tinacms/tina-astro-starter
Note: requires react to have contextual editing, happy to take on PRs from community to improve it :)

Docs - https://tina.io/docs/frameworks/astro

**Comment by sam-moreton** [comment]
Looking at the new Tina roadmap, there doesn't appear to be any mention of this, which is a shame given the amount of demand and how perfect this would be. Tina have an early opportunity to become the go-to contextual editor in one of the fastest growing static site frameworks.

**Reply by jeffsee55** [reply]
The main blocker to visual editing with Astro is that arguably one of its best features is the ability to _not_ load JS in most use cases. Because of that, when Tina updates the data on an Astro page there's no opportunity for that data to flow through the `.astro` template. 

Having said that, we support Astro when using React and the `client:load` prop, you can see an example [here](https://github.com/jeffsee55/tina-astro/blob/main/src/pages/index.astro).

We've chatted briefly with the Astro team about what we should recommend here, but don't have a solid path forward just yet.

Any thoughts on how you'd like to see this implemented are welcome, there's no ability to conditionally add the `client:load` prop, but perhaps there are some patterns where when you're in edit mode, you're rendering a different `.astro` file.

**Reply by JS-GitRepo** [reply]
Appreciate the input and clarification on Astro/React + client:load @jeffsee55. As a new dev, I was curious about this and couldn't find much information on it, almost thinking contextual editing would not be possible in Astro based on the docs. Would it be possible to add this information to the docs for clarity?

**Comment by markjaquith** [comment]
@jeffsee55 one idea I had was configuring Astro to have `/edit` endpoints that serve a `client:load` version of the corresponding endpoint.

Has obvious downsides, but might this be a way to get zero-js public pages but React-y edit pages?

**Comment by dawaltconley** [comment]
Astro v2.6 introduced [custom client directives](https://docs.astro.build/en/reference/directives-reference/#custom-client-directives), which makes it possible to have components that only hydrate in Tina's visual editor.

Very basic working demo [here](https://github.com/dawaltconley/tina-astro) using the following directive (there's probably a better way to detect the visual editor context, though I don't know how):

```mjs
/**
 * client-directives/tina.mjs
 * @type {import('astro').ClientDirective}
 */
export default (load, opts, el) => {
  const isEditor = window.top.location.pathname.startsWith('/admin')
  if (isEditor) {
    load().then(hydrate => hydrate())
  }
}
```

This gets added to the astro config via a custom integration. Then it's possible to write Tina-friendly React components like this:

```astro
<MyComponent client:tina {...props} />
```

This is not _exactly_ zero-js, because each page needs to check whether to hydrate. But that's like <1KB of js total sent to the client (if I understand Astro's hydration strategy correctly).

~Unfortunately Astro v2 currently throws an error when importing the Tina client and requires [this tagged release](https://github.com/tinacms/tinacms/issues/3764#issuecomment-1603459568).~ This is fixed.

**Reply by KevinLanvin** [reply]
I might be missing something but it does not change anything to me. I cloned your project but there is no visual editing on posts.

**Reply by dawaltconley** [reply]
@KevinLanvin looks like I put the wrong link in the readme! Or it stopped redirecting. If you navigate to _this_ path you should be able to see it: [http://localhost:3000/admin/index.html#/~](http://localhost:3000/admin/index.html#/~)

Having the post editor open to a live preview page requires a `router` property in the tina config. I just updated it to do this but it's hacky because there's only one post being displayed. Normally you'll want to route dynamically by slug or something: this is just for the purposes of showing off visual editing.

```js
name: "post",
label: "Posts",
path: "content/posts",
ui: {
  allowedActions: {
    create: false,
    delete: false,
  },
  router: () => '/',
},
```

**Reply by eerison** [reply]
Hey @dawaltconley 

I am following your steps, But it is not working for me :(

I am using [astro-paper theme](https://github.com/satnaing/astro-paper), But I couldn't make this work :(

I am trying to apply your changes on this file: https://github.com/satnaing/astro-paper/blob/main/src/pages/posts/%5Bslug%5D/index.astro

**Comment by totalhack** [comment]
The last approach from @dawaltconley seems to be similar to how CloudCannon achieves this with their bookshop integration for live editing. I'm here though because I'd like to give Tina a try, but would prefer Astro. Are there any near term plans to officially support Astro?

**Reply by wicksipedia** [reply]
This is in the backlog atm - https://github.com/tinacms/tinacms/issues/4497

We’re trying to start working on it in the next Sprint 🤞

**Reply by RebornGeek** [reply]
Any update on when this might get picked up?

**Reply by mrienstra** [reply]
There was some recent progress, see https://github.com/tinacms/tinacms/issues/5413 (completed Jan 12) and https://github.com/tinacms/tinacms/issues/5414 (Jan 27). Docs on setup with or without starter template at https://tina.io/docs/frameworks/astro

Edit: and https://github.com/tinacms/tinacms/issues/4497 was officially closed as completed, nice!

**Comment by srsgores** [comment]
It looks alright, but I think the docs need to clearly indicate whether `react` must be a dependency and used in the astro project.  Because based on what I see, for updates to happen in the editor visually, it needs react, which frankly, I don't like as a zero-JS astro user.

**Reply by wicksipedia** [reply]
👍 I'm adding it to the docs (it was already in the starter's README)

**Reply by srsgores** [reply]
I think a warning should go front and centre in the integration because a lot of people won't use the integration if they know that react is required IMO

**Reply by mrienstra** [reply]
@srsgores, I'm new here (though not to Astro) -- still evaluating -- can you help me understand the issue you have? For an end-user, the page (or post) contents will still be a `.md` file, rendered as Astro normally would, so there would be no additional JS / no performance impact, yeah? I guess it's time for me to confirm that with actual testing...

Assuming that's true and you were already thinking of it that way, your concern is... More dependencies (slower dev & build), and a slightly slower to load editing experience? Have you clocked how this compares to some other stack? [Preact](https://docs.astro.build/en/guides/integrations-guide/preact/) might mitigate some of those side effects, if it turns out to be compatible with this stack.

_(not trying to change your mind, just curious, and secondarily want to clarify for others who may run across this and be uncertain as to the scope of the problem)_

Edit: Ah, OK, so it's possible to just deploy `.md` files edited via TinaCMS without React, but doing that with the starter is far from turn-key.

**Reply by Daniel15** [reply]
@srsgores @mrienstra React isn't actually loaded on the site in production; it's only loaded when editing the site with TinaCMS. You can see this by going to the example site: https://tina-astro-starter.vercel.app/

The source code still hasl an `<astro-island>` with the serialized page data, and a rehydration script, however the rehydration script returns early (and doesn't load any of the JS) if the site isn't being edited in TinaCMS. 

(well, technically it checks [if the site is being loaded in an iframe](https://github.com/tinacms/tina-astro-starter/blob/main/astro-tina-directive/tina.js#L7), which should probably be changed to a check for TinaCMS specifically...)

**Comment by aflatter** [comment]
Hello! How well do Astro and Tina work together these days? The doc site still says [that visual editing is experimental](https://tina.io/docs/frameworks/astro#enabling-visual-editing-optional). Is that still accurate?

**Reply by Daniel15** [reply]
Visual editing works if you use React components instead of Astro components for the parts of the site that need to be editable in Tina. The production site can still be rendered to static HTML - the idea is to only rehydrate and run the JS code when the site is loaded within Tina. In prod, the JS won't be loaded other than a small snippet that checks if the site is being loaded in Tina or not. 

There's a bunch of caveats though. Since the same site is used in both prod and in Tina, the hydration data (a serialized version of all the props passed to the React components) has to be in the HTML even in prod (even though it's unused in prod), which boats out the page size quite a bit.

Visual editing is nice, but for me there's too many tradeoffs at the moment. I ended up using a different simpler CMS (Keystatic) so I could keep using Astro components and not bloat the prod HTML.

It would be great if Tina added a version of its visual editor that doesn't require React to be used. It's an understandable decision - they need some way to re-render the HTML with any arbitrary data, and React components do that well via props - but adding React for a feature that's only used to edit the site, plus a bunch of serialized JSON in the prod HTML that isn't actually used in prod, isn't great.

**Reply by JosephAlton** [reply]
@Daniel15  and others, do you see any point using Astro for making a React-heavy SSG site? I was making a site it's mainly React but a few Astro components (which could be pretty easily re-written in React).

Maybe it sounds wierd to use Astro to begin with, but I was planning on using React anyway, and a few Astro .astro components... Astro has some neat tools and plugins, such as sitemap, mdx .etc...

Just wondering whether it's worth it to use Astro or whether it means more awkwardness or boilerplate code if you are? Or just really sticking to React and trying to find replacements for the things we like in Astro with React?

Looking at the Astro starter template it looks really good, and I'm not sure why they say visual editing is not on Astro in some places....

**Reply by Daniel15** [reply]
@JosephAlton  For a site that's React-heavy, I'd probably choose something React-focused like Next.js, rather than Astro. [Next.js supports static site generation](https://nextjs.org/docs/pages/building-your-application/rendering/static-site-generation).

Astro is best suited for content-focused websites - things like blogs, marketing sites, documentation, etc. Those types of sites generally don't need a lot of JavaScript, and can mostly function with JS disabled. If they do use a library like React or Vue, might just have small sections of the site that use it. The content is the focus, and the site uses JavaScript to enhance the content, rather than JS being a core part of it. On the other hand, a site that heavily uses React might use it to render the entire page, using regular React components for client-side interactivity and React server components for static parts.

**Comment by wicksipedia** [comment]
Big news!!!

We’re getting ready to ship Astro without React. This is really going to improve the performance for all Astro users. 

We spent some time working on visual editing in Astro without React and now have TinaCMS working with Astro data islands. 

It’s currently beta and we’d love you to give it a spin.

https://github.com/tinacms/tina-astro-starter/

We’re keen to action your feedback

—- Matt Wicks, Product Owner TinaCMS
